### ISSUE TITLE
Double vs. BigDecimal for Returned API Values

### ISSUE THREAD

**Issue by zmoore-usgs** [AquaticInformatics/aquarius-sdk-java#34]
Hello,

I am a Software Developer for the U.S Geological Survey Water Mission Area and I'm currently investigating updating one of our applications - which currently directly interfaces with our Aquarius Server's API and has its own model classes - to use this SDK. Our application currently directly parses values returned from the Aquarius API into `BigDecimal` objects so that we don't run the risk of losing any amount of precision, however the SDK seems to return values as `Double` types. This possible loss of precision for values is making our team a bit hesitant to use the SDK over our own handling of returned values so I was curious if there was any possibility of considering switching the SDK over to using `BigDecimal` (or another more precise numerical type) or if there is specific reasoning behind the use of `Double`?

I appreciate any response you might have and outside of this issue the SDK is certainly looking very promising for our use case so far!

Thanks!

Sincerely,
Zackary Moore

**Comment by schmidt4brains** [comment]
Hi @zmoore-usgs, thanks for the question, and sorry for the slow response. An overly aggressive email filter ate the notification silently. I've fixed that now.

I'm less familiar with the differences between `Double` vs `BigDecimal`. so I just read up on that right now.

I chose `Double` when creating the Java SDK because it maps 1:1 with our .NET-based DTO properties (both are 64-bit IEEE 754 floating point representations).

But I see that Java offers no implicit conversion between `Double` and `BigDecimal` (probably with good reason).

Hmm. It's a bit of dilemma which way to go.

### Publish API responses

On the Publish API, I could see the value in making all the response properties `BigDecimal` instead of `Double`. Then your client code would be able to consume the responses more simply, without having to call `new BigDecimal(doubleValue)` everywhere.

Would you prefer `BigDecimal` for all properties (including things like unit group conversion factors?) or just for point values only?

### Acquisition API requests

But on the Acquisition API (append requests specifically), I think the DTO should continue to use `Double`, since that accurately reflects the precision accepted by the API.

By forcing client code to call `bigDecimalValue.doubleValue()`, it is a bit more obvious that the client code is responsible for any precision loss when converting from `BigDecimal` to `Double`.

## Does your team prefer access by public field or by public getter?

When you consume an API response, do you access the public field via `response.PropertyName`?
Or do you access the getter via `response.getPropertyName()`?

The generated DTO classes support both. Maybe there is some way to offer both flavours by added setter overrides and `getPropertyNameAsDouble()` or `getPropertyNameAsBigDecimal()`? (Welcome to Java verbosity!)

Let me know what you think.

Cheers,
Doug Schmidt

**Comment by zmoore-usgs** [comment]
Hi @schmidt4brains, thanks for the response and no worries about the delay! We've been focusing on other work until now anyways.

Yeah, the differences are a little nuanced but since we've been using BigDecimal from the start that's the direction we'd like to continue in. As far as I know the current "best" way of converting a Double to a BigDecimal without losing precision is to call BigDecimal.valueof(Double) which properly converts the value. Doing a direct call to BigDecimal(Double) creates a BigDecimal off of the rounded, exact value of the Double, rather than the value it's meant to represent. (See the second (non-accepted) answer here for more details: https://stackoverflow.com/questions/12395281/convert-double-to-bigdecimal-and-set-bigdecimal-precision).

Regarding your other questions I've talked with the team I'm working with on this project and here are our thoughts:

**Publish API**

Point values are certainly the most important values to have as BigDecimal for our use-case so we'd prefer BigDecimal for those if possible. It's probably not as important for other values as they don't tend to have the same amount of precision as point values, but having the flexibility to return them both ways as you mentioned towards the end of your post would be a welcome addition.

**Acquisition API**

I think we agree with your sentiment here about keeping these values as Double. We weren't aware that you all were using Doubles for the internal representation of these values, but knowing that now it makes sense to keep them as Double when sending them into the API.

**Public Field vs. Public Getter**

As far as the team is aware we always use getters rather than public fields so that would be the preferred direction for us. In addition getters do give the flexibility to have multiple return types as you mentioned and if that was a possibility we think having that flexibility would be a nice way to go.

Thanks for the very detailed response, hopefully we answered all of your questions. We're certainly still open to debate/discussion on these points, and we really appreciate you taking the time to look at them this in-depth!

Thanks!

Sincerely,

Zackary Moore
